src/kaleidoscope/hid.cpp

```diff
@@ -51,93 +51,225 @@
+// without (e.g. `Key_H`), which used to result in the mod flag being applied to
+// keys other than the one with the flag. By using `modifier_flag_mask`, we can
+// mask out any modifier flags that aren't attached to modifier keys or keys
+// pressed in the most recent cycle, mitigating the rollover problem, and getting
+// the intended `The` instead of `THe`.
+
+// requested_modifiers is bitmap of the modifiers attached to any non-modifier
+// key found to be pressed during the most recent cycle. For example, it would
+// include modifiers attached to Key_A, but not modifiers attached to
+// Key_LeftControl
+
+static uint8_t requested_modifiers{0};
+
+// last_keycode_toggled_on is the keycode of the key most recently toggled on
+// for this report.  This is set when a keypress is first detected and cleared 
+// after the report is sent. If multiple keys are toggled on during a single
+// cycle, this contains the most recently handled one. 
+
+static uint8_t last_keycode_toggled_on{0};
+
+void resetModifierTracking(void) {
+  last_keycode_toggled_on = 0;
+  requested_modifiers = 0;
 }
 
-void pressRawKey(Key mappedKey) {
-  WITH_BOOTKEYBOARD_PROTOCOL {
-    BootKeyboard.press(mappedKey.keyCode);
-    return;
-  }
+// isModifierKey takes a Key and returns true if the key is a 
+// keyboard key corresponding to a modifier like Control, Alt or Shift
+// TODO: This function should be lifted to the Kaleidoscope core, somewhere.
+
+bool isModifierKey(Key key) {
+  // If it's not a keyboard key, return false
+  if (key.flags & (SYNTHETIC | RESERVED)) return false;
+
+  return (key.keyCode >= HID_KEYBOARD_FIRST_MODIFIER &&
+          key.keyCode <= HID_KEYBOARD_LAST_MODIFIER);
+}
+
+// requestModifiers takes a bitmap of modifiers that might apply
+// to the next USB HID report and adds them to a bitmap of all such modifiers.
 
-  Keyboard.press(mappedKey.keyCode);
+void requestModifiers(byte flags) {
+  requested_modifiers |= flags;
 }
 
-void pressKey(Key mappedKey) {
-  if (mappedKey.flags & SHIFT_HELD) {
+// pressModifiers takes a bitmap of modifier keys that must be included in
+// the upcoming USB HID report and passes them through to KeyboardioHID
+// immediately
+
+void pressModifiers(byte flags) {
+  if (flags & SHIFT_HELD) {
     pressRawKey(Key_LeftShift);
   }
-  if (mappedKey.flags & CTRL_HELD) {
+  if (flags & CTRL_HELD) {
     pressRawKey(Key_LeftControl);
   }
-  if (mappedKey.flags & LALT_HELD) {
+  if (flags & LALT_HELD) {
     pressRawKey(Key_LeftAlt);
   }
-  if (mappedKey.flags & RALT_HELD) {
+  if (flags & RALT_HELD) {
     pressRawKey(Key_RightAlt);
   }
-  if (mappedKey.flags & GUI_HELD) {
+  if (flags & GUI_HELD) {
     pressRawKey(Key_LeftGui);
   }
+}
+
+}  // namespace
+
+
+
+void initializeKeyboard() {
+  Keyboard.begin();
+  WITH_BOOTKEYBOARD {
+    BootKeyboard.begin();
+  }
+}
+
+// pressKey takes a Key, as well as optional boolean 'toggledOn' which defaults
+// to 'true'
```

Comment:
Does this match the implementation?

Reply:
The definition of the API is over here: https://github.com/keyboardio/Kaleidoscope/pull/347/files#diff-dfa257efaf4a9a7ccbb4b63d30e4ba54R12 

I -think- it makes more sense to set the default there, but maybe a more experienced C++ dev can set me straight if I'm wrong.